target/ppc/fpu_helper.c

```diff
@@ -3339,3 +3339,173 @@ void helper_xssubqp(CPUPPCState *env, uint32_t opcode,
     *xt = t;
     do_float_check_status(env, GETPC());
 }
+
+static inline target_ulong compute_fp_flags(CPUPPCState *env)
+{
+    float_status *excp_ptr;
+    int excp_flags;
+    target_ulong fpscr_flags, fpscr;
+    excp_ptr = &env->fp_status;
+    excp_flags = get_float_exception_flags(excp_ptr);
+    fpscr_flags = 0;
+    fpscr = env->fpscr;
+    if (unlikely(excp_flags & float_flag_overflow)) {
+        fpscr_flags |= FP_FX;
+        fpscr_flags |= FP_OX;
+        fpscr_flags |= (fpscr & FP_OE) ? FP_FEX : 0;
+    }
+    if (unlikely(excp_flags & float_flag_underflow)) {
+        fpscr_flags |= FP_FX;
+        fpscr_flags |= FP_UX;
+        fpscr_flags |= (fpscr & FP_UE) ? FP_FEX : 0;
+    }
+    if (unlikely(excp_flags & float_flag_inexact)) {
+        fpscr_flags |= FP_FX;
+        fpscr_flags |= FP_XX;
+        fpscr_flags |= (fpscr & FP_XE) ? FP_FEX : 0;
+    }
+    if (unlikely(excp_flags & float_flag_invalid)) {
```

Comment:
It's a bit different here. The description of these MMA instructions says that 

> For every multiply-add operation that is performed as part of the execution of this instruction, the operation is performed as if all exception enable bits are zero, and the trap-disabled result is returned. [...] Exception status is accumulated and the appropriate exception status bits in the FPSCR are updated at the completion of execution of the instruction.

IIUC, some multiply-add operation may set inexact and other may set invalid, and both bits would be set in FPSCR. 

However, considering that softfpu accumulates `float_exception_flags` (see `float_raise` implementation in `include/fpu/softfloat.h`), I wonder if we need this method at all. Shouldn't it be enough to move the call to `helper_reset_fpstatus` outside the loop and call `do_float_check_status` at the end of the helper?

Reply:
> However, considering that softfpu accumulates float_exception_flags (see float_raise implementation in include/fpu/softfloat.h), I wonder if we need this method at all. Shouldn't it be enough to move the call to helper_reset_fpstatus outside the loop and call do_float_check_status at the end of the helper?

I don't think calling `do_float_check_status` is enough because it doesn't deal with `invalid` exceptions now (only overflow/underflow and inexact).
But I think moving the call outside the loop can still be done creating a temporary `float_status` instead of a temporary `fpscr`.

Reply:
> I don't think calling `do_float_check_status` is enough because it doesn't deal with `invalid` exceptions now (only overflow/underflow and inexact).

Oh, right, we'd need to call something like `float_invalid_op_madd` too.

> But I think moving the call outside the loop can still be done creating a temporary `float_status` instead of a temporary `fpscr`. (might not even be necessary, would need to check if any of the `float32_*` relies on the status value being cleared from the previous iteration)

AFAICT, the only place in softfpu that checks `float_exception_flags` is the hardfloat test `can_use_fpu`

Reply:
I thought it also checked more, but having a quick look at the code it seems only `float32_is_quiet_nan` and `float32_is_signaling_nan` checks `float_exception_flags`, I'll move the check outside the loop